### PR TITLE
Add workspace alert to sprite lab when sprites reach max

### PR DESCRIPTION
**Pull request by fisher-alice**
- This is a follow-up task to [this PR](https://github.com/code-dot-org/code-dot-org/pull/47738) which limits the number of sprites created in Sprite Lab to prevent the browser from freezing. Once this PR is approved, it will be merged with https://github.com/code-dot-org/code-dot-org/pull/47953 and then both parts will be merged to staging together.
- This PR displays a workspace alert warning when the user creates more than MAX_NUM_SPRITE_WARNING (999) sprites either during preview mode (auto-run) or when the program is running. The workspace alert disappears if the number of sprites created is MAX_NUM_SPRITE_WARNING or less during preview mode or when the program is reset.
- The MAX_NUM_SPRITES is set to 1000. When number of sprites is MAX_NUM_SPRITES, functions that create sprites return early which drastically improves performance.
- This PR builds on [this update](https://github.com/code-dot-org/code-dot-org/pull/47815) which added the workspace alert to the CodeWorkspace component that can be used in both App lab and Sprite lab.

## After update videos

- This video shows the workspace alert warning being displayed and closed for the function addSprite:

https://user-images.githubusercontent.com/107423305/188748396-a7688ac8-882c-47b0-a9ca-ee4054aff7c0.mp4

- This video shows the workspace alert warning being displayed and closed for the function addNumSprites:

https://user-images.githubusercontent.com/107423305/188748415-f17179aa-9cf0-4e77-9cbc-62087f454949.mp4

- This video shows the workspace alert warning being displayed and closed for the function addBursts:

https://user-images.githubusercontent.com/107423305/188748472-ef898a26-4a16-497b-944a-49e34b07e721.mp4

- This video shows the workspace alert warning being displayed and closed for two different adding sprite functions. It also shows the result when the number of iterations is very large.

https://user-images.githubusercontent.com/107423305/188748492-6ef93dfb-4d16-46a8-8e28-7e8a431a85d0.mp4

## Links
- [jira ticket - [Sprite Lab] Display workspace alert when user reaches limit](https://codedotorg.atlassian.net/browse/STAR-2376)
- [jira ticket - Prototype: How can we allow users to block the auto-run functionality in Sprite Lab](https://codedotorg.atlassian.net/browse/STAR-2323)

## Testing story
- In CoreLibraryTest.js added a unit test to check that number of sprites are limited when addSprite is called within a loop with large number of iterations. Also added a unit test to ensure a workspace alert is displayed after MAX_NUM_SPRITE_WARNING sprites has been created.

## Follow-up work
In Sprite Lab, there are additional ways to cause the browser to freeze
- A user can add behaviors to sprites. The following program will not cause the browser to freeze with this PR update:

Screen Shot 2022-08-31 at 12 22 03 PM

Although it doesn't necessarily make sense to do so, If a user adds a sprite behavior to a loop with a large number of iterations (>10,000), this will cause the browser to freeze. 

Screen Shot 2022-08-31 at 12 22 42 PM

- Using the print function within a loop will also cause the browser to freeze for large number of iterations (>500).

Both of these issues should be addressed:
- [jira - Large number of print statements cause browser freezing](https://codedotorg.atlassian.net/browse/STAR-2395)
- [jira - Large number of sprite behaviors cause browser freezing](https://codedotorg.atlassian.net/browse/STAR-2396) 

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/i18n/common/en_us.json
Comment (sanchitmalhotra126) [inline apps/i18n/common/en_us.json]: It would be good to make the 1000 in this string configurable in case we decide to alter the limit later. If you include something like `{limit}` instead of the 1000 (ex. `...number of sprites in Spite Lab is {limit}. No additional...`), then when referencing this string, you can pass in an object specifying the value of "limit", ex. `msg.spriteLimitExceeded({limit: MAX_NUM_SPRITES})`

File: apps/src/p5lab/redux/textConsole.js
Comment (sanchitmalhotra126) [inline apps/src/p5lab/redux/textConsole.js]: Should we store this 999 in constants as well? (or store it as a 1000 and convert it to -999 here so the actual limit value is clearer)

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (sanchitmalhotra126) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: nit: maybe `hasReachedSpriteLimit` or `reachedSpriteLimit` since this returns a boolean?

File: apps/test/unit/p5lab/spritelab/CoreLibraryTest.js
Comment (sanchitmalhotra126) [inline apps/test/unit/p5lab/spritelab/CoreLibraryTest.js]: nit: if you want to make this test future-proof in case MAX_NUM_SPRITES changes, you can convert the test name to a template string and replace the 1001 with ${MAX_NUM_SPRITES + 1} 🙂

**Comment by jmkulwik** [comment]
It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits.
If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically

File: apps/test/unit/p5lab/spritelab/CoreLibraryTest.js
Comment (sanchitmalhotra126) [inline apps/test/unit/p5lab/spritelab/CoreLibraryTest.js]: nit: `let`

File: apps/test/unit/p5lab/spritelab/CoreLibraryTest.js
Comment (sanchitmalhotra126) [inline apps/test/unit/p5lab/spritelab/CoreLibraryTest.js]: nit: could use MAX_NUM_SPRITES + 1 in place of 1001 here

**Review by sanchitmalhotra126 — COMMENTED** [review]
Nice! Just a few comments

File: apps/src/p5lab/redux/textConsole.js
Comment (fisher-alice) [reply apps/src/p5lab/redux/textConsole.js]: Yes - this makes sense! I updated in the prior PR that first made this change. https://github.com/code-dot-org/code-dot-org/pull/47738

**Comment by fisher-alice** [comment]
> It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits. If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically

Thanks Jessie! I merged the `sprite_lab_limit_sprites/texts` with staging to create this branch since I wanted to also access the updates I made from https://github.com/code-dot-org/code-dot-org/pull/47815 so if I change the base branch to `sprite_lab_limit_sprites/texts`, I get many more changes than the ones I've made. 

I merged the prior PR so now the diff should be more clear. Thanks!

File: apps/i18n/common/en_us.json
Comment (fisher-alice) [reply apps/i18n/common/en_us.json]: That's really cool! Thanks - will update.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Makes sense! Will use reachedSpriteLimit.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (pablo-code-org) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: small nit :) 

numSprites > MAX_NUM_SPRITES 

In my head is more human readable and matches your method definition without an operation going through the devs head.

Comment (pablo-code-org) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: sorry, my bad, 

I get confused with the definition of Max_NUM_SPRITES which is not the real max 
`(MAX_NUM_SPRITES + 1) is the actual maximum number of sprites possible`

Comment (pablo-code-org) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Can a new variable be created 

MAX_NUM_SPRITES = 1001;
MAX_NUM_SPRITE_WARNING = MAX_NUMBER_SPRITES - 1;

And then it would avoid confusions when reading the code? We could even add some comments around the second variable to understand more about it. I can see someone forgetting a +1 somewhere and introducing a bug

Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: I LOVE this idea of using two constants. I was struggling with which one to use, but much better to use 2 different constants! And I like the name MAX_NUM_SPRITE_WARNING - will use. Thanks so much!

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: For my own understanding: Do we have the `=== MAX_NUM_SPRITE_WARNING` so the warning gets displayed only once per time the user goes over the max number of sprites? i.e. if the user puts 1005, the warning is displayed once instead of 5x?

Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Yes, that's correct.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: This function reads like a boolean function, but has a side effect of triggering a warning on the code editor. This can lead to confusion when maintaining the code because someone in the future might copy what looks like a simple boolean function and get unintended side effects.

This can be fixed by splitting this into two functions. Something like `reachedSpriteLimit` and `warnIfAtSpriteLimit`.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (jmkulwik) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: (i.e. as opposed to using `>= MAX_NUM_SPRITES`)

File: apps/src/p5lab/spritelab/constants.js
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/constants.js]: I'm concerned that the MAX_NUM_SPRITES and MAX_NUM_SPRITE_WARNING could lead to future off-by-one errors that are difficult to catch. A future code maintainer would need to be very careful that they're using the right variable in the right place. Since 1000 is a somewhat arbitrary limit (i.e. choosing 1010 or 995 would lead to no real difference in user experience), just using `MAX_NUM_SPRITES = 1000` and limiting the user if `spritesRequested > MAX_NUM_SPRITES` and displaying the warning if `spritesRequested === MAX_NUM_SPRITES` seems very reasonable to me. Then there's only one variable that needs to be used everywhere and it's a little more clear to the reader what is happening in each place. The user _could_ create 1001 sprites, but the intended functionality still works. However, with the two competing variables, I worry that a future code maintainer may accidentally break the functionality and allow users to create as many sprites as they want.

I'm curious to hear others thoughts, however. This is how I would approach it, but I could see an argument in a different direction.

File: apps/test/unit/p5lab/spritelab/CoreLibraryTest.js
Comment (jmkulwik) [inline apps/test/unit/p5lab/spritelab/CoreLibraryTest.js]: NIT: should `calledWith` here be `calledOnceWith`? ([sinon-chai cheat sheet](https://www.chaijs.com/plugins/sinon-chai/) if it's helpful)

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (jmkulwik) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: (oops I had the page cached and didn't see your response. Ignore my second comment)

**Comment by mikeharv** [comment]
> > It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits. If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically
> 
> Thanks Jessie! I merged the `sprite_lab_limit_sprites/texts` with staging since I wanted to also access the updates I made from #47815 so if I change the base branch to `sprite_lab_limit_sprites/texts`, I get many more changes than the ones I've made.
> 
> I merged the prior PR so now the diff should be more clear. Thanks!

Hey Alice, I think we had talked about not merging https://github.com/code-dot-org/code-dot-org/pull/47738/ until this one was ready to go, that way we don't release a feature that only has half of what we want. I'd suggest reverting https://github.com/code-dot-org/code-dot-org/pull/47738/ and merging that revert ASAP. 

Once that's done, you can set up another PR to "revert the revert". Jessie's suggestion was to update the base branch at the top of this page to point to the original changes, but still unmerged. (See screenshot)

image

That should just make it easier to compare the changes here without also having to make sense of the changes from the original PR. Once this one is approved, you'd want to merge this one _first_. 

I hope this is somewhat clear, and I apologize if it's not! Thanks for all the great work.

**Comment by fisher-alice** [comment]
> > > It looks like this branch is based on `sprite_lab_limit_sprites/texts` from https://github.com/code-dot-org/code-dot-org/pull/47738/commits. If that's the case, go ahead and update the base branch on the pull request to `sprite_lab_limit_sprites/texts` for ease of reviewing. You can do this by clicking the "edit" button to the right of the PR title and changing the base branch from staging. That also makes it clearer to future developers what code is changed in this PR specifically
> > 
> > 
> > Thanks Jessie! I merged the `sprite_lab_limit_sprites/texts` with staging since I wanted to also access the updates I made from #47815 so if I change the base branch to `sprite_lab_limit_sprites/texts`, I get many more changes than the ones I've made.
> > I merged the prior PR so now the diff should be more clear. Thanks!
> 
> Hey Alice, I think we had talked about not merging #47738 until this one was ready to go, that way we don't release a feature that only has half of what we want. I'd suggest reverting #47738 and merging that revert ASAP.
> 
> Once that's done, you can set up another PR to "revert the revert". Jessie's suggestion was to update the base branch at the top of this page to point to the original changes, but still unmerged. (See screenshot)
> 
> image
> 
> That should just make it easier to compare the changes here without also having to make sense of the changes from the original PR. Once this one is approved, you'd want to merge this one _first_.
> 
> I hope this is somewhat clear, and I apologize if it's not! Thanks for all the great work.

Thanks Mike! I did try Jessie's suggestion to update the base branch at first. However, since I merged this branch with the latest version of staging to pull in my changes from https://github.com/code-dot-org/code-dot-org/pull/47815, the diff included all the other changes made to staging since I last pulled (yikes!). So was attempting to make it easier for Jessie to review this PR compared to the #47738.
She since was able to review so I reverted #47738 and plan to merge both PRs at the same time. Thanks so much!
Update: I reverted and differences seem to be maintained correctly :)

File: apps/test/unit/p5lab/spritelab/CoreLibraryTest.js
Comment (fisher-alice) [reply apps/test/unit/p5lab/spritelab/CoreLibraryTest.js]: Thanks! Made the update.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Thanks Jessie! This is great feedback.
I updated reachedSpriteLimit so it just checks if the total number of sprites >= MAX_NUM_SPRITES.
Then added a warnIfAtSpriteLimit to dispatch a warning if the total number of sprites == SPRITE_LIMIT_WARNING. I renamed the constant to reflect its purpose more clearly. 
I think that separating these 2 functions also clarifies the purpose of the 2 constants as well to address your comment below about the use of 2 constants instead of 1. Lmk what you think of this update.

I added a couple more unit tests to check to make sure the alert is dispatched correctly - a test to check warning is dispatched only once even when the addSprites function is called additional times past the SPRITE_LIMIT_WARNING. And then a test to make sure the alert is not dispatched at the edge case.

Thanks!

File: apps/src/p5lab/spritelab/constants.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/constants.js]: Yes - I would also like to hear other folks's thoughts as well!
I renamed the MAX_NUM_SPRITE_WARNING to SPRITE_LIMIT_WARNING and made other updates in its use (created a 2nd function) and tests. More details in my reply to comment above. Thanks!

File: apps/test/unit/p5lab/spritelab/CoreLibraryTest.js
Comment (fisher-alice) [reply apps/test/unit/p5lab/spritelab/CoreLibraryTest.js]: Also added 2 more unit tests.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (sanchitmalhotra126) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: nit: you could probably just inline this? `return this.getNumberOfSprites() >= MAX_NUM_SPRITES`

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (sanchitmalhotra126) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: nit: same here as well (`if (this.getNumberOfSprites() === SPRITE_LIMIT_WARNING)`)

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (sanchitmalhotra126) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: Minor nit, but since this set of code is repeated a few times, I don't think it needs to be commented this heavily in each case - more potential for one comment to get out of sync with others which could lead to confusion down the road. As long as the original functions are well commented (which `warnIfAtSpriteLimit` seems to be; may want to add a quick comment to `reachedSpriteMax`, though it is pretty self-explanatory 🙂) I don't think you need to include too many comments here.

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks good to me! Thanks for working through the branches and updating the PR! Just a few minor nits.

**Review by mikeharv — APPROVED** [review]
Fantastic work Alice! Thanks for all of the persistence on this one. :)

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Thanks! Made the update.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Thanks! Made the update.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Thanks - makes sense to remove repetitive comments. Made the updates!

File: apps/src/p5lab/spritelab/constants.js
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/constants.js]: NIT: When I see "SPRITE_LIMIT_WARNING" I tend to think the variable contains a string. Something such as "SPRITE_LIMIT_THRESHOLD" might be a clearer indicator that the variable contains a number.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (jmkulwik) [inline apps/src/p5lab/spritelab/CoreLibrary.js]: NIT: (I don't feel strongly about this. Feel free to ignore if you disagree.) Is `reachedSpriteMax` and `warnIfAtSpriteLimit` linked? i.e. you'd generally expect that the two commands should come one after the other? If so, I'd recommend something like this:
```
    if (this.reachedSpriteMax()) {
      return;
    } else if (this.atSpriteLimit()) {
      this.warn();
    }
```
The reasoning: future code maintainers will read an `if` clause followed by a function as two separate logical blocks but will read an `if`/`else if` clause as one logical block. The two are less likely to be separated and it's less likely that one will be deleted while the other retained.

**Review by jmkulwik — APPROVED** [review]
Great work on this!!

File: apps/src/p5lab/spritelab/constants.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/constants.js]: Great idea - thanks! I'll use SPRITE_WARNING_THRESHOLD.

File: apps/src/p5lab/spritelab/CoreLibrary.js
Comment (fisher-alice) [reply apps/src/p5lab/spritelab/CoreLibrary.js]: Thank you! I appreciate your point about future code maintainers and added a helper function similar to reachedSpriteMax called reachedSpriteWarningThreshold. Also renamed the function that dispatches the warning - I think it does make it easier to read with the parallel logic:
    if (this.reachedSpriteMax()) {
      return;
    } else if (this.reachedSpriteWarningThreshold()) {
      this.dispatchSpriteLimitWarning();
    }

File: apps/src/p5lab/spritelab/constants.js
Comment (breville) [reply apps/src/p5lab/spritelab/constants.js]: I agree with @jmkulwik that it seems a bit odd to export two numbers, one of which is always one less than the other, when you really only need to export one of them, and the business logic elsewhere can handle the warning when you are exactly one before.  However, there is one argument for keeping them as separate, and that would be if we might want to adjust this to warn at a notably different threshold than the maximum.  If that's never going to happen, though, then it seems more robust to just export one magic number.

Comment (fisher-alice) [reply apps/src/p5lab/spritelab/constants.js]: Thanks @breville ! This makes and I'm definitely open to refactor.

**Pull request merged**
